### ISSUE TITLE
Question about Hea

### ISSUE THREAD

**Issue by falcaopetri** [lajus/amie#35]
Hi,

The Head Coverage has been defined as _"the ratio of instantiations of the head atom that are predicted by the rule"_ (AMIE 3's paper). It's formula is given as:

<img src="https://render.githubusercontent.com/render/math?math=hc(\mathbf{B} \Rightarrow r(x, y)) = \frac{supp(\mathbf{B} \Rightarrow r(x, y))}{|\{(x, y) : r(x, y) \in K\}|}"> (Equation 1) 

Now, consider a rule which predicts a constant: `B => ?x relation C`. Since y is a constant, I thought that HC would be calculated as:

<img src="https://render.githubusercontent.com/render/math?math=hc(\mathbf{B} \Rightarrow r(x, C)) = \frac{supp(\mathbf{B} \Rightarrow r(x, C))}{|\{x : r(x, C) \in K\}|}"> (Equation 2) 

As I understood from code, HC uses [MiningAssistant#getHeadCardinality](https://github.com/lajus/amie/blob/a2beeca2969202602d48f421ba682acc99a02ac2/mining/src/main/java/amie/mining/assistant/MiningAssistant.java#L462-L464) which considers a fixed support for each relation. Therefore, HC is:

<img src="https://render.githubusercontent.com/render/math?math=hc(\mathbf{B} \Rightarrow r(x, C)) = \frac{supp(\mathbf{B} \Rightarrow r(x, C))}{|\{(x, y) : r(x, y) \in K\}|}">  (Equation 3) 

Also, as I understood from `InstantiatedHeadMiningAssistant` (which counts on one variable), HC would be:
<img src="https://render.githubusercontent.com/render/math?math=hc(\mathbf{B} \Rightarrow r(x, C)) = \frac{supp(\mathbf{B} \Rightarrow r(x, C))}{|\{x : r(x, y) \in K\}|}">  (Equation 4) 

1. Are these observations correct?
2. Does Equation 2 make sense? As I see, Equation 2 can be considered as Equation 3 conditioned to y=C. Therefore, both have applications, depending on how we read the associated metrics.

Regards,
Antonio.

**Comment by lajus** [comment]
Hi,

Head Coverage can be related to the notion of recall of a rule and computes how well the predictions of a rule cover the set of fact with the same head relation. In this sense, the proper formulae are Equation 1 and Equation 3.

Equation 2 computes how well the predictions of your rule, the `x`s such as `r(x, C)` cover the facts of the form `r(x, C)` of the KB. This new measure also captures a more specific form of recall and as such should "make sense", but I would not consider it as the Head Coverage.

Indeed, AMIE supposes that the head coverage is monotonous, e.g the head coverage of a refinement of a rule must be inferior than the head coverage of the initial rule. If we consider the rule Equation 3 as a refinement of the rule Equation 1 and we change the formula used, this property is not guaranteed to hold.

TL;DR: Equation 3 is not Head Coverage but a new interesting measure.

Cheers,
Jonathan

**Comment by falcaopetri** [comment]
> If we consider the rule **Equation 3** as a refinement of the rule Equation 1 and we change the formula used, this property is not guaranteed to hold.
> 
> TL;DR: **Equation 3** is not Head Coverage but a new interesting measure.

_I'm assuming that this last 2 references for "Equation 3" should read "Equation 2", right?_

--------------
Thanks for the explanation, @lajus. I forgot to consider that the monotonous property was required for pruning.

I really liked your definitions in AMIE 3's paper. My only note is that Equation 1 (from my first comment) led me to falsely assume that all *y*'s would be instantiated. This was not the case in the previous papers' definitions since they differentiated the denominator's *y'*. Of course, this difference only applies to instantiated heads.

Regards,
Antonio.

**Comment by lajus** [comment]
> 
> 
> > If we consider the rule **Equation 3** as a refinement of the rule Equation 1 and we change the formula used, this property is not guaranteed to hold.
> > TL;DR: **Equation 3** is not Head Coverage but a new interesting measure.
> 
> _I'm assuming that this last 2 references for "Equation 3" should read "Equation 2", right?_
Yes, that is right.

Jonathan
